core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -271,56 +247,82 @@ private boolean unsubscribe(String queryName,
-        SimpleQueryUpdateEmitter<I, U> emitter = new SimpleQueryUpdateEmitter<>(updateHandler);
-        I initialResult = interceptAndInvoke(uow, m -> queryHandler.handle(m, emitter)).getPayload();
-        emitter.initialResult(initialResult);
         return () -> {
-            emitter.cancelRegistration();
+            updateHandlers.remove(query);
             return true;
         };
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public <U> void emit(Predicate<SubscriptionQueryMessage<?, ?, U>> filter,
+                         Function<SubscriptionQueryMessage<?, ?, U>, U> update) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(sqm -> filter.test((SubscriptionQueryMessage<?, ?, U>) sqm))
+                              .map(m -> (SubscriptionQueryMessage<?, ?, U>) m)
+                              .forEach(query -> {
+                                  UpdateHandler<?, U> updateHandler = (UpdateHandler<?, U>) updateHandlers.get(query);
+                                  try {
+                                      U calculatedUpdate = update.apply(query);
+                                      updateHandler.onUpdate(calculatedUpdate);
+                                  } catch (Exception e) {
+                                      updateHandlers.remove(query);
+                                      updateHandler.onCompletedExceptionally(e);
+                                  }
+                              }));
+    }
+
+    @Override
+    public void complete(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(filter)
+                              .forEach(query -> {
+                                  UpdateHandler<?, ?> updateHandler = updateHandlers.get(query);
+                                  updateHandlers.remove(query);
+                                  updateHandler.onCompleted();
+                              }));
+    }
+
+    @Override
+    public void completeExceptionally(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter, Throwable cause) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(filter)
+                              .forEach(query -> {
+                                  UpdateHandler<?, ?> updateHandler = updateHandlers.get(query);
+                                  updateHandlers.remove(query);
+                                  updateHandler.onCompletedExceptionally(cause);
+                              }));
+    }
+
+    @SuppressWarnings("unchecked")
+    private void registeringSubscriptionQueryReadSafe(Predicate<?> filter, Runnable r) {
```

Comment:
You are right, but this is something which will execute the `r` which is read safe regarding subscription queries which are currently registering. I'm open for better naming here, since it causes confusion already :)

Reply:
As I pointed out I cant come up with a better name atm, hence I hope javadoc might solve any misunderstanding of this functions goal.